scripts/calculate_perplexity.py

```diff
@@ -0,0 +1,427 @@
+        )
+
+        del outputs
+
+    return loss_list
+
+
+# 3. Perplexity Calculation
+def calculate_perplexity(
+    data_loader, inference_session, ctx_len, prompt_len, model_type, batch_size, log_file, model_name
+):
+    loss_fct = torch.nn.CrossEntropyLoss(reduction="none")
+    total_loss = 0
+    total_tokens = 0
+    with open(log_file, "w") as fp:
+        cnt = 0
+
+    is_qaic = False
+    is_torch = False
+    if isinstance(model_type, str) and model_type == "qpc":
+        is_qaic = True
+    elif isinstance(model_type, str) and model_type == "torch":
+        is_torch = True
+
+    if is_torch:
+        loss_list = torch_perplexity(
+            data_loader, batch_size, inference_session, loss_fct, log_file, cnt, prompt_len, ctx_len, model_name
+        )
+        avg_loss = torch.stack(loss_list).mean()
+        perplexity = np.exp(avg_loss)
+        return perplexity, avg_loss
+
+    else:
+        for inp_ids, attn_mask in tqdm(data_loader):
+            loop_s = time.time()
+            pad_input_batch = None
+            if inp_ids.shape[0] != batch_size and attn_mask.shape[0] != batch_size:
+                pad_input_batch = batch_size - inp_ids.shape[0]
+                inp_ids = np.concatenate((inp_ids, np.repeat(inp_ids[-1:], pad_input_batch, axis=0)))
+                attn_mask = np.concatenate((attn_mask, np.repeat(attn_mask[-1:], pad_input_batch, axis=0)))
+
+            inputs = {}
+            targets_label = []
+            outputs_logits = []
+            input_ids = inp_ids[:, :prompt_len]
+            attention_mask = attn_mask[:, :prompt_len]
+            input_len = attention_mask.sum(1, keepdims=True)
+            padded_len = input_ids.shape[1]
+
+            inputs["input_ids"] = input_ids
+            inputs["position_ids"] = np.where(attention_mask, np.arange(padded_len), -1)
+            cache_index = np.array([0])
+
+            # Run inference
+            if is_qaic:
+                outputs = inference_session.run(inputs)
+                if len(outputs["logits"].shape) == 2:
+                    outputs["logits"] = np.expand_dims(outputs["logits"], axis=1)
+                if outputs["logits"].dtype == np.float16:
+                    outputs["logits"] = outputs["logits"].astype(np.float32)
+                outputs_logits.append(torch.tensor(outputs["logits"]))
+
+                cache_index += 1
+                inputs["input_ids"] = inp_ids[:, cache_index]
+                targets_label.append(torch.from_numpy(inputs["input_ids"]))
+                inputs["position_ids"] = input_len
+            else:
+                outputs = {}
+                input_names, input_shapes, output_names = inference_session._input_output_specs()
+                for i, iname in enumerate(input_names):
+                    if "past" in iname:
+                        inputs[iname] = np.zeros(
```

Comment:
For models with 3D KV layout (like starcoder1 architecture) one more case need to be added.

Reply:
@quic-akuruvil , can you push this change on the same PR pls.